### ISSUE TITLE
support vector parameters

### ISSUE THREAD

**Issue by pkienzle** [bumps/bumps#73]
See discussion in https://github.com/reflectometry/refl1d/pull/132

**Comment by bmaranville** [comment]
@hoogerheide what is missing from bumps?  Is your refl1d code not working with bumps as-is?

**Comment by hoogerheide** [comment]
Currently allowing lists of parameters is specific to the refl1d FunctionalLayer object. I believe @pkienzle  added this because it could be a general feature of Bumps parameters that lists of such parameters are allowed.

**Comment by bmaranville** [comment]
Trying to prioritize: is this a bumps feature you'd like to use in the near future, or can it be deferred to be part of the 'long term development plan'?  I.e. does this need to be in the next release, or are you cool for now?

**Comment by hoogerheide** [comment]
This is definitely a long-term planning option. Incorporating this functionality into the refl1d flayer module is sufficient for now.

**Comment by bmaranville** [comment]
I am worried it's a bit fragile to treat every parameter as potentially a list of parameters - we should discuss whether we can instead create a ParameterList or ParameterArray class that would meet the needs, and then the receiving code will know what to do with it.

**Comment by pkienzle** [comment]
Yes technical debt needs to be paid or it multiplies (working around old debt incurs additional debt).

I haven't thought about how to structure a generic API for vector parameters.

For user defined profiles we are currently accepting a function with signature `f(z: ndarray, p1: Union[float, List[float]], ...) -> ndarray`, perhaps with a default value. It does mean you need to give the an initial value list of the correct length when defining the profile since we can't change the length later, and check whether any of the parameters are lists. A type annotation would remove the guess work, but it might be harder for new python users to write.

The list of parameters translates pretty naturally into our generic parameter hierarchy, where a call to parameters() returns Tree, with `Tree = Union[None, Parameter, List["Tree"], Dict[str, "Tree"]]`.  This certainly won't scale to situations where we have enormous numbers of fit parameters, for example, if we were doing some sort of maxent analysis where we fit each pixel in an image with some constraints, but we don't have algorithms that can handle those sorts of fits.

The advantage of a ParameterList object in bumps is that the user could set properties such as range on the entire vector at once rather than looping through the list. I allow this already with parameter sets, but I doubt it is ever used. Until we get a lot of models with parameter vectors and know what might help the user work with them I think we are better off keeping it simple and leaving things as they are.

Feel free to close this ticket.
